web/track/views.py

```diff
@@ -372,51 +373,62 @@ def forgot_password():
         else:
-            # TODO: check if email exists, if it does send email
-            # response = notifications_client.send_email_notification(
-            #     email_address='email_address',
-            #     template_id='6e3368a7-0d75-47b1-b4b2-878234e554c9'
-            # )
-            # General Notification
-            msg = 'If an account is associated with the email address, ' \
-                  'further instructions will arrive in your inbox'
-            return render_template(generate_path(prefix, "forgot-password"), msg=msg)
 
-    @app.route("/en/new-password", methods=['GET', 'POST'])
-    @app.route("/fr/new-password", methods=['GET', 'POST'])
-    def new_password():
+            user_email = cleanse_input(request.form.get('email_input'))
+            user = connection.query_user_by_email(user_email)
+            return send_pass_reset(user_email, user, prefix)
+
+            # msg = 'If an account is associated with the email address, ' \
+            #     'further instructions will arrive in your inbox'
+            # return render_template(generate_path(prefix, "forgot-password"), msg=msg)
+
+    @app.route("/en/new-password/<token>", endpoint='en_new_password', methods=['GET', 'POST'])
+    @app.route("/fr/new-password/<token>", endpoint='fr_new_password', methods=['GET', 'POST'])
+    def new_password(token):
         prefix = request.path[1:3]
         if request.method == 'GET':
-            return render_template(generate_path(prefix, 'new-password'))
+            return render_template(generate_path(prefix, 'new-password'), token=token)
         else:
+            # Try to see if email matches email set in token
+            try:
+                password_reset_serial = URLSafeTimedSerializer(api_config.super_secret_key)
+                email = password_reset_serial.loads(token, salt=api_config.super_secret_salt, max_age=3600)
+            except:
+                content = error_messages.sign_in_change_pass()
+                return redirect(url_for(prefix + '_sign_in', **content))
+
+            user = connection.query_user_by_email(email)
+
             user_password = cleanse_input(request.form.get('password_input'))
             user_password_confirm = cleanse_input(request.form.get('password_confirm_input'))
 
             if user_password == user_password_confirm:
                 if is_strong_password(user_password):
                     # Create a user to insert into the database
-                    # to_add = Users(
-                    #     user_password=bcrypt.generate_password_hash(user_password),  # Flask-Bcrypt password hash
-                    # )
-                    # connection.insert(to_add)
-                    # connection.commit()
-                    return render_template(generate_path(prefix, "password-changed"))
+                    user_password = bcrypt.generate_password_hash(user_password),  # Flask-Bcrypt password hash
```

Comment:
Since our project is based on Python 3, we need to add a .decode(‘utf-8’) to the end of this hashing function call.

https://flask-bcrypt.readthedocs.io/en/latest/#usage Take a look here if you'd like!